launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+  }
+
+  protected String shQuote(String s) {
+    StringBuilder quoted = new StringBuilder();
+    boolean hasWhitespace = false;
+    for (int i = 0; i < s.length(); i++) {
+      if (Character.isWhitespace(s.codePointAt(i))) {
+        quoted.append('"');
+        hasWhitespace = true;
+        break;
+      }
+    }
+
+    for (int i = 0; i < s.length(); i++) {
+      int cp = s.codePointAt(i);
+      switch (cp) {
+        case '\'':
+          if (hasWhitespace) {
+            quoted.appendCodePoint(cp);
+            break;
+          }
+        case '"':
+        case '\\':
+          quoted.append('\\');
+          // Fall through.
+        default:
+          if (Character.isWhitespace(cp)) {
+            hasWhitespace=true;
+          }
+          quoted.appendCodePoint(cp);
+      }
+    }
+    if (hasWhitespace) {
+      quoted.append('"');
+    }
+    return quoted.toString();
+  }
+
+  // Visible for testing.
+  List<String> parseOptionString(String s) {
+    List<String> opts = new ArrayList<String>();
+    StringBuilder opt = new StringBuilder();
+    boolean inOpt = false;
+    boolean inSingleQuote = false;
+    boolean inDoubleQuote = false;
+    boolean escapeNext = false;
+    boolean hasData = false;
+
+    for (int i = 0; i < s.length(); i++) {
+      int c = s.codePointAt(i);
+      if (escapeNext) {
+        if (!inOpt) {
+          inOpt = true;
+        }
+        opt.appendCodePoint(c);
+        escapeNext = false;
+      } else if (inOpt) {
+        switch (c) {
+        case '\\':
+          if (inSingleQuote) {
+            opt.appendCodePoint(c);
+          } else {
+            escapeNext = true;
+          }
+          break;
+        case '\'':
+          if (inDoubleQuote) {
+            opt.appendCodePoint(c);
+          } else {
+            inSingleQuote = !inSingleQuote;
+          }
+          break;
+        case '"':
+          if (inSingleQuote) {
+            opt.appendCodePoint(c);
+          } else {
+            inDoubleQuote = !inDoubleQuote;
+          }
+          break;
+        default:
+          if (inSingleQuote || inDoubleQuote || !Character.isWhitespace(c)) {
+            opt.appendCodePoint(c);
+          } else {
+            finishOpt(opts, opt);
+            inOpt = false;
+            hasData = false;
+          }
+        }
+      } else {
+        switch (c) {
+        case '\'':
+          inSingleQuote = true;
+          inOpt = true;
+          hasData = true;
+          break;
+        case '"':
+          inDoubleQuote = true;
+          inOpt = true;
+          hasData = true;
+          break;
+        case '\\':
+          escapeNext = true;
+          break;
+        default:
+          if (!Character.isWhitespace(c)) {
+            inOpt = true;
+            opt.appendCodePoint(c);
+          }
+        }
+      }
+    }
+
+    checkArgument(!inSingleQuote && !inDoubleQuote && !escapeNext, "Invalid option string: %s", s);
+    if (opt.length() > 0 || hasData) {
```

Comment:
do we need to check `hasData` here? Would the behavior be the same if we just checked for `opt.length() > 0` here?